src/main/java/com/gooddata/http/client/GoodDataHttpClient.java

```diff
@@ -94,35 +92,72 @@
 
     private final SSTRetrievalStrategy sstStrategy;
 
-    private final HttpContext context;
+    /** Host performing authentication - eg. issuing TT tokens */
+    private final HttpHost authHost;
+
+    /** this lock is used to ensure that no threads will try to send requests while authentication is performed */
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+
+    /** this lock guards that only one thread enters the authentication (obtaining TT/SST) section */
+    private final Lock authLock = new ReentrantLock();
+
+    /** current SST (or null if not yet obtained) */
+    private String sst;
+
+    /** TT to be set into the header (or null if not yet obtained) */
+    private String tt;
+
 
     /**
      * Construct object.
      * @param httpClient Http client
      * @param sstStrategy super-secure token (SST) obtaining strategy
      */
+    @Deprecated
     public GoodDataHttpClient(final HttpClient httpClient, final SSTRetrievalStrategy sstStrategy) {
+        notNull(httpClient);
         this.httpClient = httpClient;
-        this.sstStrategy = sstStrategy;
-        context = new BasicHttpContext();
-        final CookieStore cookieStore = new BasicCookieStore();
-        context.setAttribute(HttpClientContext.COOKIE_STORE, cookieStore);
-
-        //this lock is used to ensure that no threads will try to send requests while authentication is performed
-        context.setAttribute(LOCK_RW, new ReentrantReadWriteLock());
-
-        //this lock guards that only one thread enters the authentication (obtaining TT/SST) section
-        context.setAttribute(LOCK_AUTH, new ReentrantLock());
+        if (sstStrategy instanceof LoginSSTRetrievalStrategy) {
+            this.sstStrategy = sstStrategy;
+            this.authHost = ((LoginSSTRetrievalStrategy) sstStrategy).getHttpHost();
+        } else {
+            throw new IllegalArgumentException("This constructor is deprecated and works with LoginSSTRetrievalStrategy argument only!");
+        }
     }
 
     /**
      * Construct object.
      * @param sstStrategy super-secure token (SST) obtaining strategy
      */
+    @Deprecated
     public GoodDataHttpClient(final SSTRetrievalStrategy sstStrategy) {
         this(HttpClientBuilder.create().build(), sstStrategy);
     }
 
+    /**
+     * Construct object.
+     * @param httpClient Http client
+     * @param authHost http host
+     * @param sstStrategy super-secure token (SST) obtaining strategy
+     */
+    public GoodDataHttpClient(final HttpClient httpClient, final HttpHost authHost, final SSTRetrievalStrategy sstStrategy) {
```

Comment:
What about introduce builder (on GoodDataHttpClient.create() ) instead of introducing another 2 new constructors (with builder we can avoid future breaking of the API & be consistent with Apache HttpClient builder style)